### PR TITLE
[notification] New method: `channel/can-connect?`

### PR DESCRIPTION
**Pull request by qnkhuat**
Part of milestone 2.1 of https://github.com/metabase/metabase/issues/43822
This PR does 2 things:
- add a new `channel/can-connect?` multimethod and implement for http
- the `channel/send!` method now has the first arg is a map with a `:type` key.

This is needed so that we can introduce a new entity: `channel`.

### CODE REVIEW

File: src/metabase/channel/http.clj [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                 (cond-> request
+                   (= :request-body auth-method) (update :body merge auth-info)
+                   (= :header auth-method)       (update :headers merge auth-info)
+                   (= :query-param auth-method)  (update :query-params merge auth-info)))))
+
+(defmethod channel/can-connect? :channel/http
+  [_channel-type details]
+  (channel.shared/validate-channel-details HTTPDetails details)
+  (try
+    (channel/send! (merge {:type :channel/http} details) {})
```
Comment (piranha) [inline src/metabase/channel/http.clj:43]: Would be nice to send a test with some test data actually so they can see what is going on on their end. With the fields we promise to be non-optional.

But I think it should be passed to `can-connect?` somewhere from the top, not generated in there. Optionally we can leave this for later if you want. :)

**Review by sansolovyov — APPROVED** [review]
Nice! A few non-blocking comments.

**Comment by qnkhuat** [comment]
@piranha, thanks for the review before it was even ready :D I added a PR description now if you're still interested.

File: src/metabase/channel/http.clj (hunk 1 above)
Comment (qnkhuat) [reply src/metabase/channel/http.clj:43]: I have a similar concern here: https://metaboat.slack.com/archives/C07318NQ5M2/p1719804756972109?thread_ts=1719018252.065269&cid=C07318NQ5M2

I think it actually makes more sense if we separate the endpoint we used to test the connection vs the endpoint we use to send the payload. but as Arakaki said, it's follow-up work.

File: src/metabase/channel/core.clj [hunk 2]
```diff
@@ -7,12 +7,29 @@
...
 (set! *warn-on-reflection* true)
 
 ;; ------------------------------------------------------------------------------------------------;;
 ;;                                      Channels methods                                           ;;
 ;; ------------------------------------------------------------------------------------------------;;
 
+(defmulti can-connect?
+  "Check whether we can connect to a `channel` with `detail`.
+
+  Returns `true` if can connect to the channel, otherwise return faslsy or throw an appriopriate exception.
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/channel/core.clj]: fixed 1f4ecac494

File: src/metabase/channel/core.clj [hunk 3]
```diff
@@ -7,12 +7,29 @@
...
 
+
 (set! *warn-on-reflection* true)
 
 ;; ------------------------------------------------------------------------------------------------;;
 ;;                                      Channels methods                                           ;;
 ;; ------------------------------------------------------------------------------------------------;;
 
+(defmulti can-connect?
+  "Check whether we can connect to a `channel` with `detail`.
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/channel/core.clj]: yes 1f4ecac494

File: src/metabase/channel/http.clj (hunk 1 above)
Comment (piranha) [reply src/metabase/channel/http.clj:43]: I don't think we need to do any authentication calls, IMO "put the token in channel config" is perfectly valid.

Github sends this when you just configure webhook: https://docs.github.com/en/webhooks/webhook-events-and-payloads#ping

**Pull request merged**